op-supervisor/supervisor/backend/syncnode/reset.go

```diff
@@ -0,0 +1,187 @@
+package syncnode
+
+import (
+	"context"
+	"errors"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+)
+
+// ongoingReset represents a bisection,
+// where a is the earliest in the range,
+// and z is the latest in the range.
+// when a and z are directly adjacent, the
+// bisection is complete. a is the last consistent
+// block, and z is the first inconsistent block.
+type ongoingReset struct {
+	a eth.BlockID
+	z eth.BlockID
+}
+
+func (m *ManagedNode) checkConsistencyWithDB() {
+	ctx, cancel := context.WithTimeout(m.ctx, internalTimeout)
+	defer cancel()
+
+	var z eth.BlockID
+
+	// check if the last unsafe block we saw is consistent with the logs db
+	err := m.backend.IsLocalUnsafe(ctx, m.chainID, m.lastNodeLocalUnsafe)
+	if errors.Is(err, types.ErrConflict) {
+		m.log.Warn("local unsafe block is inconsistent with logs db. Initiating reset",
+			"lastUnsafeblock", m.lastNodeLocalUnsafe,
+			"err", err)
+		z = m.lastNodeLocalUnsafe
+	}
+
+	// check if the last safe block we saw is consistent with the local safe db
+	err = m.backend.IsLocalSafe(ctx, m.chainID, m.lastNodeLocalSafe)
+	if errors.Is(err, types.ErrConflict) {
+		m.log.Warn("local safe block is inconsistent with logs db. Initiating reset",
+			"lastSafeblock", m.lastNodeLocalSafe,
+			"err", err)
+		z = m.lastNodeLocalSafe
+	}
+
+	// there is inconsistency. initiate reset
+	if z != (eth.BlockID{}) {
+		m.resetting.Store(true)
+		m.cancelReset.Store(false)
+		m.ongoingReset = &ongoingReset{a: eth.BlockID{}, z: z}
+		// trigger the reset asynchronously
+		go m.prepareReset()
+	} else {
+		m.log.Debug("no inconsistency found")
+		m.resetting.Store(false)
+		m.ongoingReset = nil
+	}
+}
+
+// prepareReset prepares the reset by bisecting the the search range
+// until the last consistent block is found. It then identifies the correct
+// unsafe, safe, and finalized blocks to target for the reset.
+func (m *ManagedNode) prepareReset() {
+	internalCtx, iCancel := context.WithTimeout(m.ctx, internalTimeout)
+	defer iCancel()
+	nodeCtx, nCancel := context.WithTimeout(m.ctx, nodeTimeout)
+	defer nCancel()
```

Comment:
We should apply these contexts for each individual request, not for all of them. If the bisection is 20+ layers deep (longer chain history), and the connection is slow or rate-limited for some reason, it might time out otherwise.